### PR TITLE
Add unread_threads tag to templatetags

### PR DESCRIPTION
**Pull request by mattlindesay**
Additional template tag to show if there are new messages for the user so that you can highlight this on a navbar.

### CODE REVIEW

**Comment by mattlindesay** [comment]
Oops. Closed by mistake.

**Comment by mattlindesay** [comment]
OK no idea why this is not compiling on django master. I don't think it's the code I added. If someone can explain why it's failing I'm happy to spend some time on this.

**Comment by grahamu** [comment]
@mattlindesay recently I have begun removing "master" from our tox.ini and .travis.yml files, replacing them with "1.10". The reason is two-fold. Django dev master no longer allows "from django.core.urlresolvers import reverse", requiring "from django.urls import ..." instead. Secondly, Pinax is tested against "master" mainly to stay on top of the latest changes, but doesn't officially support latest dev. So my suggestion is to remove "master" from the test matrix configurations in both tox.ini and .travis.yml and try again.

**Comment by grahamu** [comment]
See [pinax-blog tox.ini](https://github.com/pinax/pinax-blog/blob/master/tox.ini) for an example of this "master" removal and "1.10" insertion.

File: pinax/messages/templatetags/pinax_messages_tags.py [hunk 1]
```diff
@@ -7,9 +7,15 @@
...
+    '''
+    Check whether there are any unread messages for a particular thread for a user.
+    '''
     return bool(thread.userthread_set.filter(user=user, unread=True))
 
 
 @register.filter
 def unread_threads(user):
+    '''
+    Return the number of unread threads for this user, useful for highlighting on an account bar for example.
```
Comment (grahamu) [inline pinax/messages/templatetags/pinax_messages_tags.py]: "Return the number of Threads with unread messages for this user, ..."

**Comment by grahamu** [comment]
Hi again @mattlindesay. Looks like removing "master" and adding "1.10" fixed the build, and adding a test fixed coveralls. Nice work.

Just three small issues remain for this to get merged in:

* Since you're looking for a count of unread Threads rather than a list of actual Threads, I think the Django doc [optimization guidance](https://docs.djangoproject.com/en/1.10/topics/db/optimization/#don-t-retrieve-things-you-don-t-need) is good; use [count()](https://docs.djangoproject.com/en/1.10/topics/db/optimization/#use-queryset-count).

* If I understand correctly, this filter produces a count of Threads which have one or more messages not read by user. Proving this works as intended is good to a) show proper use, and b) show other developers what to expect. You've added one test but I think this needs a bit more clarity. Edge cases are good, so maybe three Threads, one unread entirely, one with two messages - one read and another not read, and a third Thread with all messages read by user. Ensure the templatetag produces expected result.

* Somehow all file modes have changed, 100644 → 100755. All other Pinax apps use 644, and I think that's standard convention.

Thanks for contributing Matt!

File: pinax/messages/tests/tests.py [hunk 2]
```diff
@@ -233,6 +233,72 @@ def test_unread(self):
...
+            [self.jtauber],
+            "Why did you break the internet?", "I demand to know.").thread
+
+        tmpl = """
+               {% load pinax_messages_tags %}
+               {% if user|unread_threads %}{{ user|unread_threads }}{% endif %}
+               """
+        self.assert_renders(
+            tmpl,
+            Context({"thread": thread, "user": self.jtauber}),
```
Comment (grahamu) [inline pinax/messages/tests/tests.py]: Context does not need "thread" variable. Simplify and clarify by removing "thread" in this and other tests.

File: pinax/messages/tests/tests.py [hunk 3]
```diff
@@ -233,6 +233,72 @@ def test_unread(self):
...
+            self.brosner,
+            [self.jtauber],
+            "Second message", "So there are two.").thread
+        tmpl = """
+               {% load pinax_messages_tags %}
+               {% if user|unread_threads %}{{ user|unread_threads }}{% endif %}
+               """
+        self.assert_renders(
+            tmpl,
+            Context({"thread": thread, "user": self.jtauber}),
```
Comment (grahamu) [inline pinax/messages/tests/tests.py]: remove Context "thread" as noted above

File: pinax/messages/tests/tests.py [hunk 4]
```diff
@@ -233,6 +233,72 @@ def test_unread(self):
...
+
+        Message.new_reply(thread, self.jtauber, "Replying to the message so that I have no unread Threads")
+
+        tmpl = """
+               {% load pinax_messages_tags %}
+               {{ user|unread_threads }}
+               """
+        self.assert_renders(
+            tmpl,
+            Context({"thread": thread, "user": self.jtauber}),
```
Comment (grahamu) [inline pinax/messages/tests/tests.py]: remove Context "thread" as noted above

File: pinax/messages/tests/tests.py [hunk 5]
```diff
@@ -233,6 +233,72 @@ def test_unread(self):
...
+        Ensure `unread_threads` template_tag produces correct results for one unread message
+        """
+        thread = Message.new_message(
+            self.brosner,
+            [self.jtauber],
+            "Why did you break the internet?", "I demand to know.").thread
+
+        tmpl = """
+               {% load pinax_messages_tags %}
+               {% if user|unread_threads %}{{ user|unread_threads }}{% endif %}
```
Comment (grahamu) [inline pinax/messages/tests/tests.py]: Replace with cleaner `{{ user|unread_threads }}`, no point in the conditional for these tests.

File: pinax/messages/tests/tests.py [hunk 6]
```diff
@@ -233,6 +233,72 @@ def test_unread(self):
...
+        Message.new_reply(thread, self.jtauber, "Replying to the first message")
+        Message.new_reply(thread, self.brosner, "Replying again, so that there are two unread messages on one thread")
+
+        thread = Message.new_message(
+            self.brosner,
+            [self.jtauber],
+            "Second message", "So there are two.").thread
+        tmpl = """
+               {% load pinax_messages_tags %}
+               {% if user|unread_threads %}{{ user|unread_threads }}{% endif %}
```
Comment (grahamu) [inline pinax/messages/tests/tests.py]: replace with `{{ user|unread_threads }}` as noted above

File: pinax/messages/tests/tests.py [hunk 7]
```diff
@@ -233,6 +233,72 @@ def test_unread(self):
...
+               """
+        self.assert_renders(
+            tmpl,
+            Context({"thread": thread, "user": self.jtauber}),
+            '2'
+        )
```
Comment (grahamu) [inline pinax/messages/tests/tests.py]: I think it would be smart to also check the number of Threads with unread messages for "brosner" in this test. I believe the value should be "1".

File: docs/templatetags.md [hunk 8]
```diff
@@ -16,5 +16,23 @@ For instance if there are unread messages in a thread, change the CSS class acco
     </div>
 ```
 
+## unread_threads
+
+Returns the number of unread threads for the user. Use for notifying a user of new messages, for example in _account_bar.html
```
Comment (grahamu) [inline docs/templatetags.md]: "Returns the number of Threads with unread messages for the user...

**Comment by grahamu** [comment]
BTW @mattlindesay you (and anyone else) are welcome to join Pinax discussions in the various channels of https://pinax.slack.com.

**Comment by grahamu** [comment]
Thanks for your contribution @mattlindesay. I've incorporated this improvement into pinax-messages via https://github.com/pinax/pinax-messages/pull/42, so closing this pull request.
